src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonInEnumerationCheck.java

```diff
@@ -0,0 +1,116 @@
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2019 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * <p>
+ * Checks if unnecessary semicolon is in enum definitions.
+ * Semicolon is not needed if enum body is empty.
+ * The check will flag the following with warnings:
+ * </p>
+ * <pre>
+ * enum One {
+ *     A,B; // violation
+ * }
+ * enum Two {
+ *     A,B,; // violation
+ * }
+ * enum Three {
+ *     A,B(); // violation
+ * }
+ * enum Four {
+ *     A,B{}; // violation
+ * }
+ * enum NoSemicolon {
+ *     A,B // OK
+ * }
+ * enum Normal {
+ *     A,
+ *     B
+ *     ; // required ";", no violation
+ *     Normal(){}
+ * }
+ * </pre>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;UnnecessarySemicolonInEnumeration&quot;/&gt;
+ * </pre>
+ *
+ * @since 8.22
+ */
+@StatelessCheck
+public final class UnnecessarySemicolonInEnumerationCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_SEMI = "unnecessary.semicolon";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.ENUM_CONSTANT_DEF,
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final DetailAST semicolon = getSemicolonSibling(ast);
+        if (semicolon != null && hasCodeElementsAfter(semicolon)) {
+            log(semicolon, MSG_SEMI);
+        }
+    }
+
+    private static DetailAST getSemicolonSibling(DetailAST ast) {
```

Comment:
how about
```
    private static DetailAST findSemicolonSibling(DetailAST ast) {
        DetailAST curr = ast;
        do {
            curr = curr.getNextSibling();
        } while (curr != null && curr.getType() != TokenTypes.SEMI);
        return curr;
    }
```
?

Reply:
It looks better, but there is one flaw - if there is no semicolon, it will return some other token instead of semicolon token. 
I think it reasonable to return null if we found no semicolon.

Reply:
>  it will return some other token instead of semicolon token.

No. The `do-while` loop may ends only when
* there is no more siblings (at this time  `curr=null`)
* there is a semicolon (`curr.getType==TokenTypes.SEMI`)

Reply:
I should have drunk more coffee before writing my last comment.

Yes, you are right.
Thanks, corrected.